compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -401,18 +403,35 @@ impl<'alloc> Candidates<'alloc> {
         // We use `retain` here to remove the elements from the reverse set if we've removed the
         // matching candidate in the forward set.
         srcs.retain(|src| {
-            if !f(*src) {
+            if f(*src) == CandidateModification::NoChange {
                 return true;
             }
             let Entry::Occupied(entry) = self.c.entry(*src) else {
                 return false;
             };
-            Self::entry_remove(entry, *src, |dest| dest == p, at);
+            Self::entry_modify_candidates(
+                entry,
+                *src,
+                |dest| {
+                    if dest == p {
+                        CandidateModification::Remove
+                    } else {
+                        CandidateModification::NoChange
+                    }
+                },
+                at,
+            );
             false
         });
     }
 }
 
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum CandidateModification {
+    NoChange,
+    Remove,
+}
```

Comment:
The modification sounds very general while the operation is limited to just filtering. Maybe `enum Filter { Keep, Remove }`, while using `filter` in method names?